lib/src/puppetdb-cli.cc

```diff
@@ -116,6 +209,57 @@ size_t write_body(char *ptr, size_t size, size_t nmemb, void *userdata){
     return size * nmemb;
 }
 
+// SynchronizedCharQueueWrapper is a rapidjson adapter which satisfies the
+// `ReadStream` interface for a rapidjson::Reader. This allows us to parse data
+// from curl as we recieve it.
+class SynchronizedCharQueueWrapper {
+  public:
+    typedef char Ch;
+    SynchronizedCharQueueWrapper(SynchronizedCharQueue& stream) : stream_(stream) {}
+    char Peek() const {
+        int c = Front();
+        return c == char_traits<char>::eof() ? '\0' : (char)c;
+    }
+    char Take() {
+        int c = Get();
+        return c == char_traits<char>::eof() ? '\0' : (char)c;
+    }
+    size_t Tell() const { return (size_t)stream_.size(); }
+    char* PutBegin() { assert(false); return 0; }
+    void Put(char c) { assert(false); }
+    void Flush() { assert(false); }
+    size_t PutEnd(char* c) { assert(false); return 0; }
+  private:
+    int Front() const { return stream_.front(); }
+    int Get() { return stream_.pop(); }
+    SynchronizedCharQueueWrapper(const SynchronizedCharQueueWrapper&);
+    SynchronizedCharQueueWrapper& operator=(const SynchronizedCharQueueWrapper&);
+    SynchronizedCharQueue& stream_;
+};
+
+void write_pretty(UserData& userdata) {
+    // This will make sure to block until we have parsed the header
+    userdata.stream.front();
+
+    boost::string_ref content_type{ userdata.content_type };
+    if (content_type.starts_with("application/json")) {
+        rapidjson::Reader reader;
+        SynchronizedCharQueueWrapper is(userdata.stream);
+        char writeBuffer[65536];
```

Comment:
Sorry, didn't see this ping earlier.  Facter [just wraps std::cout with a rapidjson adapter](https://github.com/puppetlabs/facter/blob/master/lib/src/facts/collection.cc#L477-L495) that writes each character given it (not the most efficient thing in terms of operator<< calls, but the ostream will buffer the actual underlying I/O operations at least).

Doing it this way is fine, but it probably doesn't need a buffer this large.